### PR TITLE
Avoid negative integer overflow when `filesize < io_->tell()`.

### LINKED ISSUE #791: Integer overflow, leading to very large allocation in WebPImage::decodeChunks()
There is a bug at [webpimage.cpp:515](https://github.com/Exiv2/exiv2/blob/55dfdb9ab007d6e6d1d06240048690eaac23b305/src/webpimage.cpp#L515). If `filesize < io_->tell()` then the subtraction overflows and the bounds check on `size` is ineffective.

There is an additional problem in this function, which is that the error conditions are not checked after the calls to `io_->read`. Calls to `io_->read` should always be followed by calls to `io_->error()` and `io_->eof()`, to make sure that the buffer was read successfully. This means that the code does not notice that the file size is much smaller than the value of `size` would suggest. This has enabled me to write a PoC which causes exiv2 to run for almost a minute, scanning 4GB of uninitialized memory.

Here is the PoC, which I have tested on master ([55dfdb9](https://github.com/Exiv2/exiv2/commit/55dfdb9ab007d6e6d1d06240048690eaac23b305)): 
poc1. You can run the PoC like this to see the long-running behavior:

```
$ time exiv2 poc1.jpg  
File name       : poc1.jpg
File size       : 28 Bytes
MIME type       : image/webp
Image size      : 0 x 0
poc1.jpg: No Exif data found in the file

real    0m52.738s
user    0m50.621s
sys     0m2.108s
```

Or if you use `ulimit` to lower the maximum allocation size to less than 4GB then exiv2 will crash:

```
$ ulimit -v 4000000
$ exiv2 poc1.jpg
Uncaught exception: std::bad_alloc
```

### PR DESCRIPTION
**Pull request by kevinbackhouse**
This fixes #791 on the master branch.

It is mostly the same as the fix for v0.27: #797.

This fix is nicer because I have added a new method named `readOrThrow` to `BasicIo`. But I couldn't make the same change in v0.27 because it changes the API of `BasicIo`.

### CODE REVIEW

**Comment by piponazo** [comment]
Hi @kevinbackhouse. Just in case you did not notice, the windows build is broken after these changes:

```
..\src\webpimage.cpp(515): error C2220: warning treated as error - no 'object' file generated
..\src\webpimage.cpp(515): warning C4018: '<=': signed/unsigned mismatch
```

Do you have a windows development environment where to check/fix this? Otherwise I could take a look.

File: src/webpimage.cpp [hunk 1]
```diff
@@ -508,10 +509,11 @@ namespace Exiv2 {
 #endif
 
         chunkId.pData_[4] = '\0' ;
-        while ( !io_->eof() && (uint64_t) io_->tell() < filesize) {
+        while (!io_->eof() && io_->tell() < filesize) {
             io_->readOrThrow(chunkId.pData_, WEBP_TAG_SIZE, Exiv2::kerCorruptedMetadata);
             io_->readOrThrow(size_buff, WEBP_TAG_SIZE, Exiv2::kerCorruptedMetadata);
-            const uint32_t size = Exiv2::getULong(size_buff, littleEndian);
+            const long size = Exiv2::getULong(size_buff, littleEndian);
+            enforce(0 <= size, Exiv2::kerCorruptedMetadata);
```
Comment (D4N) [inline src/webpimage.cpp]: This check relies on undefined behavior (signed integer overflow) and is thus not effective, as the compiler is free to optimize it away.

File: src/webpimage.cpp [hunk 2]
```diff
@@ -489,15 +489,16 @@ namespace Exiv2 {
...
         chunkId.pData_[4] = '\0' ;
 
-        io_->read(data, WEBP_TAG_SIZE * 3);
+        io_->readOrThrow(data, WEBP_TAG_SIZE * 3, Exiv2::kerCorruptedMetadata);
 
-        const uint32_t filesize = Exiv2::getULong(data + WEBP_TAG_SIZE, littleEndian) + 8;
-        enforce(filesize <= io_->size(), Exiv2::kerCorruptedMetadata);
+        const long filesize = Exiv2::getULong(data + WEBP_TAG_SIZE, littleEndian) + 8;
+        enforce(0 <= filesize, Exiv2::kerCorruptedMetadata);
+        enforce((size_t)filesize <= io_->size(), Exiv2::kerCorruptedMetadata);
```
Comment (D4N) [inline src/webpimage.cpp]:
```suggestion
        enforce(static_cast<size_t>(filesize) <= io_->size(), Exiv2::kerCorruptedMetadata);
```

File: src/webpimage.cpp [hunk 3]
```diff
@@ -489,15 +489,16 @@ namespace Exiv2 {
         DataBuf chunkId(5);
         chunkId.pData_[4] = '\0' ;
 
-        io_->read(data, WEBP_TAG_SIZE * 3);
+        io_->readOrThrow(data, WEBP_TAG_SIZE * 3, Exiv2::kerCorruptedMetadata);
 
-        const uint32_t filesize = Exiv2::getULong(data + WEBP_TAG_SIZE, littleEndian) + 8;
-        enforce(filesize <= io_->size(), Exiv2::kerCorruptedMetadata);
+        const long filesize = Exiv2::getULong(data + WEBP_TAG_SIZE, littleEndian) + 8;
```
Comment (D4N) [inline src/webpimage.cpp]: Add a `#include "safe_op.hpp"` at the top of the file and remove the check in the following line. `Safe:add` catches overflow errors for you:
```suggestion
        const long filesize = Safe::add(Exiv2::getULong(data + WEBP_TAG_SIZE, littleEndian), 8);
```

File: src/webpimage.cpp [hunk 4]
```diff
@@ -489,15 +489,16 @@ namespace Exiv2 {
...
+        enforce(0 <= filesize, Exiv2::kerCorruptedMetadata);
```
Comment (D4N) [inline src/webpimage.cpp]:
```suggestion
```

**Review by D4N — DISMISSED** [review]
I really like your proposed new API, but I am not so sure whether I really like the current interface.

**Comment by D4N** [comment]
@kevinbackhouse Thanks for stepping forward and proposing this new API!

Since this is quite an important change, I'd like to discuss the API design first:
- we should be using `size_t` instead of `long` for the size parameter (depends on #786)
- the proposed API uses `DataBuf`, which we try to get rid off, so I'd prefer not to introduce a new API function that returns it
- I don't think the exception code should be a parameter that the API consumer can set, imho a generic read error would be fine

I've been thinking about something comparable and the best I came up with is the following:
``` C++
/// Read nbytes bytes into `dest` (enlarging it as needed)
/// @return a slice to the read chunk
[[nodiscard]] Slice<std::vector<uint8_t>> read(std::vector<uint8_t>& dest, size_t nbytes);
```

Provided you use the returned slice to access the data, this wouldn't even have to throw an exception on reaching `EOF`, as the slice will automatically throw if you perform an out of bound access. However, it might be useful to have this function "autothrow" on EOF (maybe make this configurable)?

Thoughts/opinions?

**Comment by kevinbackhouse** [comment]
@D4N: I think all the signed/unsigned casts are fixed now. (Also in #797.)

I removed the `DataBuf` API since I wasn't actually using it. I only added it for completeness, so there's no need if it's going to be deprecated soon.

I removed the `ErrorCode` parameter from the API. I replaced it with always throwing `kerInputDataReadFailed`, because that sounded the most appropriate to me for a generic error message.

Wouldn't the simplest replacement for the `DataBuf` API be something like this?

```
std::vector<uint8_t> readOrThrow(size_t nbytes);
```

Note: I don't know the history of the `DataBuf` API or the reasons why you are trying to get rid of it.

**Comment by kevinbackhouse** [comment]
I see that some of the builds are failing now due to clang's `-Wtautological-constant-out-of-range-compare` warning complaining about this line:

```
enforce(size_u32 <= static_cast<size_t>(std::numeric_limits<long>::max()), ...
```

How do you want me to fix that?

**Comment by D4N** [comment]
> @D4N: I think all the signed/unsigned casts are fixed now. (Also in #797.)

Indeed, LGTM now.

> 
> I removed the `DataBuf` API since I wasn't actually using it. I only added it for completeness, so there's no need if it's going to be deprecated soon.
> 
> I removed the `ErrorCode` parameter from the API. I replaced it with always throwing `kerInputDataReadFailed`, because that sounded the most appropriate to me for a generic error message.

Sounds good to me.

> 
> Wouldn't the simplest replacement for the `DataBuf` API be something like this?
> 
> ```
> std::vector<uint8_t> readOrThrow(size_t nbytes);
> ```

Yes, but this has the disadvantage that you always get a heap allocation and cannot start reading at an arbitrary position inside a user provided buffer (although that also doesn't work with my proposed API).

> 
> Note: I don't know the history of the `DataBuf` API or the reasons why you are trying to get rid of it.

`DataBuf` is just an inferior version of `std::vector`: it provides raw access to the underlying array, it has no automatic bounds checking (like `std::vector::at()`), it's memory management is relatively questionable, uses `long` where one should use `size_t`, etc.pp.

> I see that some of the builds are failing now due to clang's `-Wtautological-constant-out-of-range-compare` warning complaining about this line:
> 
> ```
> enforce(size_u32 <= static_cast<size_t>(std::numeric_limits<long>::max()), ...
> ```
> 
> How do you want me to fix that?

Do you want a quick fix or a proper one?

A quick fix is to include that code only on Windows, as every other platform has `sizeof(long) > sizeof(uint32_t)` and there the above code is a NOP.

A proper fix would be that I or `$volunteer` implements `Safe::cast<T>`, which has been on my todo list for quite some time.

File: src/webpimage.cpp [hunk 5]
```diff
@@ -522,7 +522,7 @@ namespace Exiv2 {
             const uint32_t size_u32 = Exiv2::getULong(size_buff, littleEndian);
 
             // Check that `size_u32` is safe to cast to `long`.
-            enforce(size_u32 <= static_cast<size_t>(std::numeric_limits<long>::max()),
+            enforce(static_cast<size_t>(size_u32) <= static_cast<size_t>(std::numeric_limits<long>::max()),
```
Comment (D4N) [inline src/webpimage.cpp]: Wrap the enforce in a 
```C++
#if defined(_WIN32) || defined(__CYGWIN__) || defined(__MINGW__) || defined(_MSC_VER)
enforce(...);
#endif
```

Comment (kevinbackhouse) [reply src/webpimage.cpp]: Thanks, I'll do that if I cannot defeat the compiler warning. I have one more thing that I will try first.

**Review by piponazo — APPROVED** [review]

**Pull request merged**